File: sdk/go/protomcp/log.go
package protomcp

import (
	pb "github.com/msilverblatt/protomcp/gen/proto/protomcp"
)

type ServerLogger struct {
	sendFn func(*pb.Envelope) error
	logger string
}

func NewServerLogger(sendFn func(*pb.Envelope) error, logger string) *ServerLogger {
	return &ServerLogger{sendFn: sendFn, logger: logger}
}

func (l *ServerLogger) log(level, dataJSON string) {
	if l.sendFn == nil {
		return
	}
	l.sendFn(&pb.Envelope{
		Msg: &pb.Envelope_Log{
			Log: &pb.LogMessage{
				Level:    level,
				Logger:   l.logger,
				DataJson: dataJSON,
			},
		},
	})
}

func (l *ServerLogger) Debug(msg string)     { l.log("debug", msg) }
func (l *ServerLogger) Info(msg string)      { l.log("info", msg) }
func (l *ServerLogger) Notice(msg string)    { l.log("notice", msg) }
func (l *ServerLogger) Warning(msg string)   { l.log("warning", msg) }
func (l *ServerLogger) Error(msg string)     { l.log("error", msg) }
func (l *ServerLogger) Critical(msg string)  { l.log("critical", msg) }
func (l *ServerLogger) Alert(msg string)     { l.log("alert", msg) }
func (l *ServerLogger) Emergency(msg string) { l.log("emergency", msg) }

File: sdk/python/src/protomcp/log.py
import json

from . import protomcp_pb2 as pb

class ServerLogger:
    def __init__(self, send_fn, name: str = ""):
        self._send_fn = send_fn
        self._name = name

    def _log(self, level: str, message: str, data=None):
        data_json = json.dumps(data) if data else json.dumps({"message": message})
        envelope = pb.Envelope(
            log=pb.LogMessage(
                level=level,
                logger=self._name,
                data_json=data_json,
            )
        )
        self._send_fn(envelope)

    def debug(self, message, **kwargs): self._log("debug", message, kwargs.get("data"))
    def info(self, message, **kwargs): self._log("info", message, kwargs.get("data"))
    def notice(self, message, **kwargs): self._log("notice", message, kwargs.get("data"))
    def warning(self, message, **kwargs): self._log("warning", message, kwargs.get("data"))
    def error(self, message, **kwargs): self._log("error", message, kwargs.get("data"))
    def critical(self, message, **kwargs): self._log("critical", message, kwargs.get("data"))
    def alert(self, message, **kwargs): self._log("alert", message, kwargs.get("data"))
    def emergency(self, message, **kwargs): self._log("emergency", message, kwargs.get("data"))
